sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client.py

```diff
@@ -105,18 +106,18 @@ def __init__(
         self._blob_account_url = blob_account_url
 
         self._blob_service_client = BlobServiceClient(blob_account_url, credential, **kwargs)
-        self._blob_service_client._hosts[LocationMode.SECONDARY] = ""
+        self._blob_service_client._hosts[LocationMode.SECONDARY] = ""  # type: ignore [index]
```

Comment:
What is the type of `_hosts`? This ignore is in a couple of places and wondering if we can fix it.

Reply:
This is the error from `mypy`:

```
Value of type "Optional[Any]" is not indexable  [index]
```

`mypy` thinks `self.*._hosts[...] = ...` is somehow `Optional[Any]`. 

Looking into this further, in `base_client.py`, the constructor for `StorageAccountHostsMixin` initially does a

```python3
self._hosts = kwargs.get("_hosts")
```

which could be `None` (I'm not sure what the exact type of this is though). However, it later does a `None` check:

```python3
if not self._hosts:
    if len(account) > 1:
        secondary_hostname = parsed_url.netloc.replace(account[0], account[0] + "-secondary")
    if kwargs.get("secondary_hostname"):
        secondary_hostname = kwargs["secondary_hostname"]
    primary_hostname = (parsed_url.netloc + parsed_url.path).rstrip('/')
    self._hosts = {LocationMode.PRIMARY: primary_hostname, LocationMode.SECONDARY: secondary_hostname}
```
so `mypy` should theoretically know this is of `Dict[str, str]` type. I haven't found a clean way to resolve this, and like blob and file share, we decided to just do a `# type: ignore [index]` here.

Reply:
Ok, so let's actually fix this for all packages since we can. Let's add a member variable to `StorageAccountHostsMixin` like `_hosts: Dict[str, str]`.

Then change `self._hosts = kwargs.get("_hosts")` -> `self._hosts = kwargs.get("_hosts", {})` to set the default value to an empty dictionary so it's not `None`. Then I think `if not self._hosts` will still work if the dictionary is empty? If not, check for empty dict instead.

Then update all the places we set this in all SDKs to not use any ignore :)